CFOUR Calculator

bapride11 opened the pull request:
This (more appropriately tested) pull request includes:

1. A CFOUR calculator interface (only implements Calculator, not OverlapCalculator) that handles the change in frame that CFOUR does by default when using Cartesian coordinates
2. The ability to interpret an XYZ file in bohr via including the word 'bohr' in the comment line (akin to OpenMolcas), useful for CFOUR users who can nearly copy-paste the "QCOMP" block from CFOUR output, which is far more used to Z-matrices
3. Support for reading CFOUR Hessians (FCMFINAL) for use with IRC and the like

Sorry for the earlier, poorly tested pull request. I had dug up some old code and trusted my old notes that things were properly tested.

eljost commented on pysisyphus/calculators/CFOUR.py:
```diff
@@ -0,0 +1,191 @@
+    def __init__(
+            self,
+            cfour_input,
+            keep_molden=True,
```
Please rename `keep_molden` to `wavefunction_dump`.

eljost commented on pysisyphus/calculators/CFOUR.py:
```diff
@@ -0,0 +1,191 @@
+        self.gradient_fn = 'GRD'
+        self.to_keep = ("out.log", "density:__den.dat")
+        self.initden = None
+        if keep_molden:
```
Update to `self.wavefunction_dump`.

eljost commented on pysisyphus/calculators/CFOUR.py:
```diff
@@ -0,0 +1,191 @@
+            self.initden = kept_fns["density"]
+        except KeyError:
+            self.log("den.dat not found!")
+            return
```
`return` is not needed here.

eljost commented on pysisyphus/calculators/CFOUR.py:
```diff
@@ -0,0 +1,191 @@
+
+        return input_str
+
+    def get_energy(self, atoms, coords):
```
Signature should be `get_energy(self, atoms, coord, **prepare_kwargs)`.

eljost commented on pysisyphus/calculators/CFOUR.py:
```diff
@@ -0,0 +1,191 @@
+    def get_energy(self, atoms, coords):
+        return self.run_calculation(atoms, coords, "energy")
+
+    def get_forces(self, atoms, coords):
```
Signature should be `get_forces(self, atoms, coord, **prepare_kwargs)`.

eljost commented on pysisyphus/calculators/CFOUR.py:
```diff
@@ -0,0 +1,191 @@
+
+    def read_geom(self, text):
+        regex = r"Coordinates used in calculation \(QCOMP\)(.+)Interatomic distance matrix"
+        floats =  [self.float_regex for i in range(3)]
```
You can keep it like this. Personally I would have written `[self.float_regex] * 3`

eljost commented on pysisyphus/calculators/CFOUR.py:
```diff
@@ -0,0 +1,191 @@
+
+    def prepare(self, inp):
+        path = super().prepare(inp)
+        if self.initden:
```
I guess this is an inital guess/converged MOs from a previous calculation?

bapride11 replied:
Yes, but unfortunately it's a density rather than MO coefficients. CFOUR has two ways of feeding in guesses for the SCF, both only mildly documented:

1. Feed in a file named OLDMOS which contains MO coefficients
2. Feed in a file called initden.dat which contains an initial guess electronic density

There's a case in which initden will work and OLDMOS won't. I think it has something to do with symmetry/occupation number, but my memory is not good enough to be confident on that. I chose the more robust option in this case, but it could later be extended to also allow for OLDMOS.

eljost replied:
This is fine; I guess it doesn't matter that it is the density, because this is what a QC code constructs from a given set of MOs...

eljost reviewed (changes requested):
Dear Greg,

thanks for this PR. We actually have a CFOUR license in our group so I guess I can actually test the code, haha.

Some further remarks:
- Please update your comments and remove the double/triple ## and ###. Please replace them with single #-characters.
- Please run `black`, at least on CFOUR.py
- Try to move as much of the methods outside of the `CFOUR.py` class. Possible candidates are `rotate_gradient()`, `calc_centroid()`, `calc_rot_matrix()` and the energy/gradient parsing. All these methods don't really rely on any state of the Calculator and only require a filename or some coordinates. So these methods should be *promoted* to functions that take the relevant parameters. Please see the [ORCA.py on the current dev-branch](https://github.com/eljost/pysisyphus/blob/dev/pysisyphus/calculators/ORCA.py) to get an idea. This way the functions are more accessible and not buried in the CFOUR object.
- Please add one or two tests under the directory `tests/test_cfour`. Maybe a HF/STO-3G energy calculation on water and a short IRC run that utilizes a Hessian and runs for 3 or 4 steps in each direction.

bapride11 commented on pysisyphus/helpers.py:
```diff
@@ -51,7 +51,8 @@ def geom_from_xyz_file(xyz_fn, coord_type="cart", **coord_kwargs):
     xyz_fn = str(xyz_fn)
     atoms, coords, comment = parse_xyz_file(xyz_fn, with_comment=True)
-    coords *= ANG2BOHR
+    if 'bohr' not in comment.lower():
```
Thank you for this suggestion. I was honestly a little hesitant to add a check that broad, but had modeled it after the OpenMolcas behavior that I found useful.

CFOUR doesn't natively produce any XYZ files, but rather strings of the type:
```
 ----------------------------------------------------------------
         Coordinates used in calculation (QCOMP)
 ----------------------------------------------------------------
 Z-matrix   Atomic            Coordinates (in bohr)
  Symbol    Number           X              Y              Z
 ----------------------------------------------------------------
     O         8        -0.00000158    -0.12872748     0.00000000
     H         1         1.41821256     1.02148914     0.00000000
     H         1        -1.41818754     1.02150944     0.00000000
 ----------------------------------------------------------------
```
near the beginning of the output. It usually relies on Z-matrix input from the user (defaulting to Angstrom). Copying and pasting this (with the necessary changes to make an XYZ file) is probably the easiest way to reliably transfer coordinates to pysisyphus, given possible differences in bohr-angstrom conversion.

I really like the idea of the `IN_BOHR` token, which is a lot more specific and equally easy to remember.

bapride11 commented on pysisyphus/calculators/Calculator.py:
```diff
@@ -362,7 +362,7 @@ def prepare_turbo_coords(self, atoms, coords):
         return coord_str
 
-    def prepare_coords(self, atoms, coords):
+    def prepare_coords(self, atoms, coords, angstrom=True):
```
Good idea. I ended up moving all this logic to be contained within CFOUR.py.

bapride11 commented:
Hi Johannes,

It's always a pleasure to read your feedback. It truly does improve the code significantly.

I think I've addressed all the concerns mentioned above with the latest series of commits. Let me know if there are any other changes you'd like before merging.

Best,
Greg

eljost commented on pysisyphus/calculators/CFOUR.py:
```diff
@@ -0,0 +1,249 @@
+        Parameters
+        ----------
+        cfour_input : dict
+            CFOUR keywords and values. Note: "on" must be encapsulated in quotes to avoid being translated to True by YAML.
```
Did you run `black` on this file? At least this line seems quite long; or does `black` not format long comments?!
Please add a linebreak at a sensible position.

bapride11 replied:
I did run black, but apparently it doesn't format long comments. I'm as surprised as you are. Fixing.

eljost commented on pysisyphus/calculators/CFOUR.py:
```diff
@@ -0,0 +1,249 @@
+        ----------
+        cfour_input : dict
+            CFOUR keywords and values. Note: "on" must be encapsulated in quotes to avoid being translated to True by YAML.
+        keep_molden : bool, optional
```
Update to `wavefunction_dump`.

eljost commented on pysisyphus/helpers.py:
```diff
@@ -51,7 +51,8 @@ def geom_from_xyz_file(xyz_fn, coord_type="cart", **coord_kwargs):
     xyz_fn = str(xyz_fn)
     atoms, coords, comment = parse_xyz_file(xyz_fn, with_comment=True)
-    coords *= ANG2BOHR
+    if "IN_BOHR" not in comment.split():
```
Please also add a comment that highlights again, that the produced `xyz`-string will usually be in Angstrom and only when `IN_BOHR` is present, it will be in Bohr.

eljost reviewed (changes requested):
LGTM for now. I'll try to run the test and see how it goes.

eljost commented:
Today, I ran my first CFOUR calculation and it appeard to me, that a `GENBAS`-file is required next to the `ZMAT` input file. `GENBAS` does not seem to appear in your calculator.
 Did I miss a keyword that convinces CFOUR to use some "built-in" `GENBAS` file or how do you handle this?

bapride11 commented:
The `GENBAS` file right now is also handled in the runscript, by symlinking the built-in GENBAS file each time right before CFOUR is run. We could try to handle this in pysisyphus, but it would either involve some new configuration key in .pysisyphusrc specifying the install path for CFOUR or requiring the CFOUR install path as part of every input file.

eljost reviewed (approved):
lgtm

eljost commented:
Dear Greg,

thanks again for this PR. The tests ran fine w/ only minor modifications of the submit-script you provided.

All the best
Johannes
